certbot-apache/certbot_apache/configurator.py

```diff
@@ -94,26 +95,49 @@ class ApacheConfigurator(augeas_configurator.AugeasConfigurator):
 
     OS_DEFAULTS = dict(
         server_root="/etc/apache2",
-        vhost_root="/etc/apache2/sites-available",
+        vhost_root="/etc/apache2/sites-enabled",
         vhost_files="*",
         logs_root="/var/log/apache2",
+        ctlpath="apache2ctl",
+        binpath="apache2",
         version_cmd=['apache2ctl', '-v'],
         apache_cmd="apache2ctl",
         restart_cmd=['apache2ctl', 'graceful'],
         conftest_cmd=['apache2ctl', 'configtest'],
         enmod=None,
         dismod=None,
         le_vhost_ext="-le-ssl.conf",
-        handle_mods=False,
+        handle_modules=False,
         handle_sites=False,
         challenge_location="/etc/apache2",
         MOD_SSL_CONF_SRC=pkg_resources.resource_filename(
             "certbot_apache", "options-ssl-apache.conf")
     )
 
-    def constant(self, key):
-        """Get constant for OS_DEFAULTS"""
-        return self.OS_DEFAULTS.get(key)
+    def option(self, key):
+        """Get a value from options"""
+        return self.options.get(key)
+
+    def _prepare_options(self):
+        """
+        Set the values possibly changed by command line parameters to
+        OS_DEFAULTS constant dictionary
+        """
+        opts = ["enmod", "dismod", "le_vhost_ext", "server_root", "vhost_root",
+                "logs_root", "challenge_location", "handle_modules", "handle_sites",
+                "ctlpath", "binpath"]
+        for o in opts:
+            # Config options use dashes instead of underscores
+            if self.conf(o.replace("_", "-")):
+                self.options[o] = self.conf(o.replace("_", "-"))
+            else:
+                self.options[o] = self.OS_DEFAULTS[o]
+
+        # Special cases
+        self.options["apache_cmd"] = self.option("ctlpath")
+        self.options["version_cmd"][0] = self.option("binpath")
```

Comment:
I did change it for consistency, the change was on purpose.

Reply:
OK just making sure.

Now that we're using a new executable, should we add a [check in prepare](https://github.com/certbot/certbot/blob/cdc333491be3cc48dd17c5191f815859035bfdc7/certbot-apache/certbot_apache/configurator.py#L205) that the executable exists (and call `path_surgery` to fix up the user's paths if not)?

Also, did you check that this works on all the different OSes being modified here?

Reply:
The check is moved to a helper method and both of the executables are now checked.

The question of all different OSes is a harder one, because of the implementation in `configurator.py` is a catch-all for all distros that don't have a matching override.

Now when rethinking this decision it might be best to use the old value to ensure compatibility.

Reply:
> Now when rethinking this decision it might be best to use the old value to ensure compatibility.

I think there's an decent argument for that. We're also now requiring that an extra executable path be configured properly when one would do before.

As I said in other comments here, I think it's a little awkward you can set `--apache-bin` but it may never be used, but it's not so bad. Unless there's a good reason to switch to the Apache binary that I haven't heard yet, I think I'd like us to:

1. Switch back to using `ctl` here.
2. Don't check for the path to `bin` unless it's going to be used.